packages/flutter/lib/src/rendering/list_wheel_viewport.dart

```diff
@@ -529,35 +529,151 @@ class RenderListWheelViewport
+      if (_createChild(index, after: null)) {
+        _layoutChild(firstChild, constraints, index);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /// Layout base on children provided by [childManager].
+  ///
+  /// From the current scroll offset, we can calculate the minimum index and
+  /// maximum index that is visible in the viewport. We can also get the index
+  /// range of the children that are currently active.
+  /// We will modify the child list so that children that are not visible
+  /// will be removed, and those that should visible but not yet created will be
+  /// instantiate by the childManager.
   @override
   void performLayout() {
-    double currentOffset = 0.0;
+    final BoxConstraints childConstraints =
+      constraints.copyWith(
+        minHeight: _itemExtent,
+        maxHeight: _itemExtent,
+        minWidth: 0.0,
+      );
+
+    // The index range that we want to spawn children.
+    int targetFirstIndex = scrollOffsetToIndex(offset.pixels - size.height / 2 - _itemExtent / 2);
+    int targetLastIndex = scrollOffsetToIndex(offset.pixels + size.height / 2 + _itemExtent / 2);
+
+    if (childManager.childSize != null) {
+      targetFirstIndex = targetFirstIndex.clamp(0, childManager.childSize - 1);
+      targetLastIndex = targetLastIndex.clamp(0, childManager.childSize - 1);
+    }
+
+    // When the render object is laying out for the first time, the child list
+    // is empty - populate it with any child in target range.
+    if (firstChild == null &&
+        !_insertAnyChild(targetFirstIndex, targetLastIndex, childConstraints))
+        return;
+
+    int currentFirstIndex = indexOf(firstChild);
+    int currentLastIndex = indexOf(lastChild);
+
+    // Remove all unnecessary children.
+    while(currentFirstIndex < targetFirstIndex) {
+      _destroyChild(firstChild);
+      currentFirstIndex++;
+    }
+
+    while(currentLastIndex > targetLastIndex) {
+      _destroyChild(lastChild);
+      currentLastIndex--;
+    }
+
+    // It can happen that after removing unnecessary children, there's no child
+    // left. In that case insert any child in target range.
+    if (firstChild == null &&
+        !_insertAnyChild(targetFirstIndex, targetLastIndex, childConstraints))
+      return;
+
+    // Relayout all active children.
     RenderBox child = firstChild;
-    final BoxConstraints innerConstraints =
-        constraints.copyWith(
-          minHeight: _itemExtent,
-          maxHeight: _itemExtent,
-          minWidth: 0.0,
-        );
-    while (child != null) {
-      child.layout(innerConstraints, parentUsesSize: true);
-      final ListWheelParentData childParentData = child.parentData;
-      // Centers the child in the cross axis. Consider making it configurable.
-      final double crossPosition = size.width / 2.0 - child.size.width / 2.0;
-      childParentData.offset = new Offset(crossPosition, currentOffset);
-      currentOffset += _itemExtent;
+    while(child != null) {
```

Comment:
nit: here and elsewhere: `while` should have a space after it. It's control flow, not a method call.